packages/kit/src/runtime/control.js

```diff
@@ -31,3 +31,28 @@ export class Redirect {
 		this.location = location;
 	}
 }
+
+export class ValidationError {
+	/**
+	 * @param {number} status
+	 * @param {Record<string, any>} errors
+	 * @param {FormData | Record<string, any> | null | undefined} [values]
+	 */
+	constructor(status, errors, values) {
+		if (values instanceof FormData) {
+			/** @type {Record<string, string>} */
+			const converted = {};
+			for (const [key, value] of values.entries()) {
+				if (typeof value === 'string') {
+					converted[key] = value;
+				}
+			}
```

Comment:
this is buggy — it won't handle this case:

```html
<form>
  <input name="x">
  <input name="x">
</form>
```

I don't think we should accept a `Record` here. It makes it seem like you can return nested fields, and fields that aren't strings, etc. Clarity over convenience, always

Reply:
We need to return JSON though since in the progressive enhancement case the values need to be transferred this way. Seperatley but related, I think it would be nicer if `$form.values` is a POJO and not `FormData`; `values.get('x')`, you can do `values.x` instead, which feels better to me.

Reply:
We need to return JSON, but it needs to include all the data, and since any field can be duplicated it means everything needs to be an array of strings even in the common case where there's a single value, and I think this...

```js
$form.values.get('thing')
```

...is probably nicer than this:

```js
$form.values.thing[0]
```

I've seen APIs that do this...

```js
for (const [key, value] of values) {
  if (converted[key]) {
    if (typeof converted[key] === 'string') {
      converted[key] = [converted[key], value];
    } else {
      converted[key].push(value);
    }
  } else {
    converted[key] = value;
  }
}
```

...and it's a disaster when you have the 'could be one, could be multiple' case. I don't know. There isn't a great answer here.

Reply:
Since we decided to allow to return arbitrary data, this is no longer relevant - resolving.